src/azure-cli/azure/cli/command_modules/storage/operations/fs_file.py

```diff
@@ -0,0 +1,143 @@
+
+
+def exists(cmd, client):
+    try:
+        client.get_file_properties()
+        return True
+    except HttpResponseError as ex:
+        from azure.cli.command_modules.storage.sdkutil import _dont_fail_on_exist
+        StorageErrorCode = cmd.get_models("_shared.models#StorageErrorCode",
+                                          resource_type=ResourceType.DATA_STORAGE_FILEDATALAKE)
+        _dont_fail_on_exist(ex, StorageErrorCode.blob_not_found)
+        return False
+
+
+def get_file_properties(client):
+    from .._transformers import transform_fs_access_output
+
+    prop = client.get_file_properties()
+    if prop.content_settings.content_md5 is not None:
+        from msrest import Serializer
+        prop.content_settings.content_md5 = Serializer.serialize_bytearray(prop.content_settings.content_md5)
+
+    acl = transform_fs_access_output(client.get_access_control())
+    result = dict(prop, **acl)
+    return result
+
+
+def list_fs_files(client, path=None, recursive=True, num_results=None, timeout=None, exclude_dir=None, marker=None):
+    generator = client.get_paths(path=path, max_results=num_results, recursive=recursive, timeout=timeout)
+    pages = generator.by_page(continuation_token=marker)
+
+    result = next(pages)
+
+    if pages.continuation_token:
+        logger.warning('Next Marker:')
+        logger.warning(pages.continuation_token)
+
+    if exclude_dir:
+        return list(f for f in result if not f.is_directory)
+
+    return result
+
+
+def upload_file(cmd, client, local_path, overwrite=None, content_settings=None, metadata=None, timeout=None,
+                if_match=None, if_none_match=None, if_modified_since=None, if_unmodified_since=None,
+                umask=None, permissions=None):
+
+    count = os.path.getsize(local_path)
+    with open(local_path, 'rb') as stream:
+        data = stream.read(count)
+    from azure.core import MatchConditions
+    upload_file_args = {
+        'content_settings': content_settings,
+        'metadata': metadata,
+        'timeout': timeout,
+        'if_modified_since': if_modified_since,
+        'if_unmodified_since': if_unmodified_since,
+    }
+
+    # Precondition Check
+    if if_match:
+        if if_match == '*':
+            upload_file_args['match_condition'] = MatchConditions.IfPresent
+        else:
+            upload_file_args['etag'] = if_match
+            upload_file_args['match_condition'] = MatchConditions.IfNotModified
+
+    if if_none_match:
+        upload_file_args['etag'] = if_none_match
+        upload_file_args['match_condition'] = MatchConditions.IfModified
+
+    # Overwrite sets to true will overwrite existing data.
+    # Overwrite sets to false will be able to upload to empty file and non-existing file,
+    # but will raise exception when uploading to non-empty file with new data.
+    if not overwrite:
+
+        # For non-existing file path, create one
+        if not exists(cmd, client):
+            client.create_file()
```

Comment:
what if overwrite is true but the file does not exist

Reply:
It will create a new file as shown here in sdk: 
https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/storage/azure-storage-file-datalake/azure/storage/filedatalake/_upload_helper.py#L52-L62